Rcpp no longer builds with RCPP_NO_SUGAR defined

This is mostly due to the fact that `RNGScope` is defined inside `inst/include/Rcpp/stats/random/random.h` and is used outside the stats directory.  There is also a small issue with the `Matrix op scalar` functions in `inst/include/Rcpp/vector/Matrix.h`.

Moving the definition of `RNGScope` outside of stats (to its own file) and including it unconditionally will fix the former.  Protecting the `Matrix op scalar` functions with `#ifndef RCPP_NO_SUGAR` will fix the latter.

Here's an example:

test.cpp:
```c++
#define RCPP_NO_SUGAR

#include <Rcpp.h>

using namespace Rcpp;

// [[Rcpp::export]]
NumericVector test_no_sugar(NumericVector v) {
        return v;
}
```

```
> library(Rcpp)
> sourceCpp("test.cpp")
test.cpp: In function ‘SEXPREC* sourceCpp_1_test_no_sugar(SEXP)’:
test.cpp:19:5: error: ‘RNGScope’ is not a member of ‘Rcpp’
     Rcpp::RNGScope rcpp_rngScope_gen;
     ^
make: *** [test.o] Error 1
g++ -I/usr/share/R/include -DNDEBUG    -I"/usr/local/lib/R/site-library/Rcpp/include" -I"/home/dcdillon/src/git"    -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c test.cpp -o test.o
/usr/lib/R/etc/Makeconf:141: recipe for target 'test.o' failed
Error in sourceCpp("test.cpp") : 
  Error 1 occurred building shared library.
> 
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -6,6 +6,7 @@
         of RNGScope
         * inst/include/Rcpp/vector/Matrix.h: Protected Matrix op Scalar
         definitions with #ifndef RCPP_NO_SUGAR
+
 2016-11-10  Dirk Eddelbuettel  <edd@debian.org>
 
         * inst/include/Rcpp/date_datetime/newDateVector.h: Added constructor

--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,12 @@
+2016-11-13  Daniel C. Dillon   <dcdillon@gmail.com>
+        * inst/include/Rcpp/RNGScope.h: Moved definition of RNGScope
+        to its own file
+        * inst/include/Rcpp.h: Unconditionally included RNGScope.h
+        * inst/include/Rcpp/stats/random/random.h: Removed definition
+        of RNGScope
+        * inst/include/Rcpp/vector/Matrix.h: Protected Matrix op Scalar
+        definitions with #ifndef RCPP_NO_SUGAR
+
 2016-11-04  Nathan Russell     <russell.nr2012@gmail.com>
 
         * inst/include/Rcpp/macros/dispatch.h: Modify variadic macros 

--- a/ChangeLog
+++ b/ChangeLog
@@ -1,4 +1,4 @@
-2016-11-13  Daniel C. Dillon   <dcdillon@gmail.com>
+2016-11-13  Daniel C. Dillon  <dcdillon@gmail.com>
         * inst/include/Rcpp/RNGScope.h: Moved definition of RNGScope
         to its own file
         * inst/include/Rcpp.h: Unconditionally included RNGScope.h

--- a/inst/NEWS.Rd
+++ b/inst/NEWS.Rd
@@ -20,6 +20,8 @@
       fixing \ghit{549})
       \item \code{Range} Sugar now used \code{R_xlen_t} type for start/end
       (PR \ghpr{568} by Qiang Kou)
+      \item Defining \code{RCPP_NO_SUGAR} no longer breaks the build.
+      (PR \ghpr{585} by Daniel C. Dillon)
     }
     \item Changes in Rcpp unit tests
     \itemize{

--- a/inst/include/Rcpp.h
+++ b/inst/include/Rcpp.h
@@ -66,6 +66,8 @@
 
 #include <Rcpp/Nullable.h>
 
+#include <Rcpp/RNGScope.h>
+
 #ifndef RCPP_NO_SUGAR
 #include <Rcpp/sugar/sugar.h>
 #include <Rcpp/stats/stats.h>

--- a/inst/include/Rcpp/RNGScope.h
+++ b/inst/include/Rcpp/RNGScope.h
@@ -1,0 +1,14 @@
+#ifndef Rcpp__RNGScope_h
+#define Rcpp__RNGScope_h
+
+namespace Rcpp {
+
+class RNGScope{
+public:
+    RNGScope(){ internal::enterRNGScope(); }
+    ~RNGScope(){ internal::exitRNGScope(); }
+};
+
+} // namespace Rcpp
+
+#endif

--- a/inst/include/Rcpp/RNGScope.h
+++ b/inst/include/Rcpp/RNGScope.h
@@ -1,3 +1,24 @@
+// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+//
+// RNGScope.h: Rcpp R/C++ interface class library --
+//
+// Copyright (C) 2010 - 2016  Douglas Bates, Dirk Eddelbuettel and Romain Francois
+//
+// This file is part of Rcpp.
+//
+// Rcpp is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// Rcpp is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Rcpp.  If not, see <http://www.gnu.org/licenses/>.
+
 #ifndef Rcpp__RNGScope_h
 #define Rcpp__RNGScope_h
 

--- a/inst/include/Rcpp/stats/random/random.h
+++ b/inst/include/Rcpp/stats/random/random.h
@@ -24,12 +24,6 @@
 
 namespace Rcpp{
 
-class RNGScope{
-public:
-    RNGScope(){ internal::enterRNGScope(); }
-    ~RNGScope(){ internal::exitRNGScope(); }
-};
-
 template <typename T>
 class Generator : public RNGScope {
 public:

--- a/inst/include/Rcpp/vector/Matrix.h
+++ b/inst/include/Rcpp/vector/Matrix.h
@@ -247,6 +247,7 @@ inline std::ostream &operator<<(std::ostream & s, const Matrix<REALSXP, StorageP
     return s;
 }
 
+#ifndef RCPP_NO_SUGAR
 #define RCPP_GENERATE_MATRIX_SCALAR_OPERATOR(__OPERATOR__)                                                                    \
     template <int RTYPE, template <class> class StoragePolicy, typename T >                                                   \
     inline typename traits::enable_if< traits::is_convertible< typename traits::remove_const_and_reference< T >::type,        \
@@ -281,6 +282,7 @@ RCPP_GENERATE_SCALAR_MATRIX_OPERATOR(*)
 RCPP_GENERATE_SCALAR_MATRIX_OPERATOR(/)
 
 #undef RCPP_GENERATE_SCALAR_MATRIX_OPERATOR
+#endif
 
 template<template <class> class StoragePolicy >
 inline std::ostream &operator<<(std::ostream & s, const Matrix<INTSXP, StoragePolicy> & rhs) {

